lib/note-detail.jsx

```diff
@@ -1,62 +1,170 @@
-import React, { PropTypes } from 'react';
-import marked from 'marked';
+import React, { Component, PropTypes } from 'react';
 import Textarea from 'react-textarea-autosize';
-import { noop, get, debounce } from 'lodash';
 import analytics from './analytics';
+import marked from 'marked';
+import { get, debounce, invoke } from 'lodash';
 import { viewExternalUrl } from './utils/url-utils';
 
-const uninitializedNoteEditor = { focus: noop };
 const saveDelay = 2000;
 
-export default React.createClass( {
+const hasModKeys = ( keys, event ) => keys.some( k => get( event, `${ k }Key` ) );
+const isValidNote = note => note && note.id;
+const prependTab = l => `\t${ l }`;
+const removeLeadingTab = l => l && l.length && l[0] === '\t' ? l.substring( 1 ) : l;
 
-	propTypes: {
+export default class NoteDetail extends Component {
+	static propTypes = {
 		note: PropTypes.object,
 		previewingMarkdown: PropTypes.bool,
 		fontSize: PropTypes.number,
 		onChangeContent: PropTypes.func.isRequired
-	},
+	};
 
-	componentWillMount: function() {
+	constructor( props ) {
+		super( props );
+
+		this.storeNoteEditor = r => this.noteEditor = r;
 		this.queueNoteSave = debounce( this.saveNote, saveDelay );
-		this.noteEditor = uninitializedNoteEditor;
-	},
+	}
 
-	componentDidMount: function() {
+	componentDidMount() {
 		// Ensures note gets saved if user abruptly quits the app
 		window.addEventListener( 'beforeunload', this.queueNoteSave.flush );
-	},
-
-	initializeNoteEditor: function( noteEditor ) {
-		this.noteEditor = noteEditor;
-	},
-
-	isValidNote: function( note ) {
-		return note && note.id;
-	},
+	}
 
-	componentWillReceiveProps: function() {
+	componentWillReceiveProps() {
 		this.queueNoteSave.flush();
-	},
+	}
 
-	componentDidUpdate: function() {
+	componentDidUpdate() {
 		const { note } = this.props;
 		const content = get( note, 'data.content', '' );
-		if ( this.isValidNote( note ) && this.noteEditor ) {
+		if ( isValidNote( note ) && this.noteEditor ) {
 			this.noteEditor.value = content;
 
 			// Let's focus the editor for new and empty notes
 			if ( content === '' ) {
-				this.noteEditor.focus();
+				invoke( this.noteEditor, 'focus' );
 			}
 		}
-	},
+	}
 
-	componentWillUnmount: function() {
+	componentWillUnmount() {
 		window.removeEventListener( 'beforeunload', this.queueNoteSave.flush );
-	},
+	}
+
+	handleKey = event => {
+		if ( ! this.noteEditor || this.props.previewingMarkdown ) {
+			return;
+		}
+
+		invoke( {
+			Tab: this.interceptTabPresses,
+			Enter: this.interceptEnterPresses,
+		}, event.key, event );
+	};
+
+	interceptEnterPresses = event => {
+		const {
+			selectionStart,
+			value,
+		} = this.noteEditor;
+
+		if ( hasModKeys( [ 'alt', 'ctrl', 'meta', 'shift' ], event ) ) {
+			return true;
+		}
+
+		const prevLineStart = value.lastIndexOf( '\n', selectionStart - 1 ) + 1;
+		const prevLine = value.substring( prevLineStart, selectionStart - 1 );
+		const prevIndent = prevLine.match( /^(\s+[-*])/ );
```

Comment:
since this is a single capturing group with the whole regexp inside it there's no need for it to be a group at all, you can remove the `(` and `)` and use `prevIndent[ 0 ]` below